### PR TITLE
fix(pop): byte stuff the body of a retrieved message

### LINKED ISSUE #96: Byte stuff the body of a retrieved message
`POPConnection.retr` streams the body of a message verbatim and then appends the terminator unconditionally.

RFC 1939 section 3 requires that a line of the body starting with a `.` is byte stuffed with a second one, so that it is not mistaken for the end of the message. Without it a client stops reading at the first such line and the rest of the message is lost.

Reproduced with a body of `b"normal line\r\n.hidden\r\n"`, which reaches the wire as:

```
+OK 20 octets
normal line
.hidden

.
```

so the client sees the message ending at `.hidden`.

The unconditional terminator is a second, smaller problem: a body that already ends with a newline gains a spurious empty line before the final dot.

The fix has to work across the boundaries of the chunks that the body is streamed in, since a line may be split between two reads, which is why it is being tracked on its own rather than bundled with the other fixes.

### PR DESCRIPTION
**Pull request by joamag**
Closes #96.

## Problem

`POPConnection.retr` streamed the body of a message verbatim and then appended the terminator unconditionally, which breaks the framing that RFC 1939 section 3 defines in two ways.

A line of the body that starts with a `.` was served as is, so a client took the first of them for the end of the message and dropped everything that followed. The body reported in the issue reached the wire as:

```
+OK 22 octets
normal line
.hidden

.
```

The unconditional terminator is the second, smaller problem, visible above as the empty line that precedes the final dot, which a body that already ends with a newline does not need.

## Changes

* A new `stuff` method duplicates the dot of every line of the body that starts with one, so that such a line is no longer taken for the terminator.
* The terminator is now preceded by a newline only when the body does not end with one, which removes the empty line that was being added.
* The connection carries a `tail` with the last bytes of the body that has been served so far, reset by `retr` at the start of every retrieval.

The `tail` is what makes the stuffing survive the chunks that the body is streamed in. It holds the last **two** bytes rather than one, because a single one is not enough to detect the case where the boundary of a chunk falls between the carriage return and the line feed of a newline. Those two bytes are prepended to the next chunk as the context of the match and then removed from the result, as they have already been served. It starts as a newline so that the first line of the body is recognised as being at the start of a line.

## Tests

Added to `src/netius/test/servers/pop.py`, ordered to mirror the declaration order of the connection so that the retrieval and the stuffing tests sit between the ones of `uidl` and `dele`. Nine of the ten fail against the code before the fix:

| Test | What it pins |
| --- | --- |
| `test_retr_stuffed` | the body reported in the issue, the dot line is stuffed |
| `test_retr_stuffed_first` | the first line of the body is a line start as well |
| `test_retr_line` | a body that ends with a newline gains no empty line |
| `test_retr_empty` | an empty body is served as the terminator alone |
| `test_retr_chunked` | the stuffing survives a newline split between two reads |
| `test_stuff`, `test_stuff_inner` | dots at the start of a line are stuffed, inner ones are not |
| `test_stuff_boundary` | the dot opens the chunk that follows the newline |
| `test_stuff_boundary_split` | the newline itself is split between the chunks |

`test_retr` covers a body with no trailing newline, which behaves the same before and after, so it stands as a guard rather than a regression test. `test_init` was added because the suite builds its connections through `__new__` and never exercised the constructor, which is where the initial value of the tail lives.

Coverage of the new and changed code:

| File | Statements in the change | Covered |
| --- | --- | --- |
| `src/netius/servers/pop.py` | 11 | 100.0% |
| `src/netius/servers/smtp.py` | 10 | 100.0% |
| **total** | **21** | **100.0%** |

No partial branches are introduced by the change. `pop.py` as a whole moves from 51.4% to 55.7%.

## Validation

Suite status: 1166 passed and 7 skipped on Linux with Python 3.14, 1165 passed and 8 skipped on macOS. `black --check` is clean over 338 files and `mypy.stubtest` reports no issues in 132 modules, the stub having been updated for the new method and attribute. Both changed files were also compiled under Python 2.7, as that interpreter is still part of the matrix.

Beyond the unit tests, the complete retrieval was driven end to end for nine bodies, including the one of the issue, an empty one, one whose first line opens with a dot, consecutive dot lines and the two chunk boundary cases. Every one of them fails against the current `master`.

## The other half, undoing the stuffing of the sender

Review of this branch surfaced that the POP stuffing alone is not enough, and that shipping it on its own would corrupt the mail that arrives through SMTP.

The ingestion path stores the wire form rather than the message. `on_raw_data` forwards the raw payload of the DATA command verbatim and `on_data_smtp` appends it as it is, with nothing undoing the stuffing that a compliant sender applies. A source line of `.hidden` is therefore stored as `..hidden`.

Before this branch the two mistakes cancelled each other out: the stuffing of the sender survived into storage and the POP server never added its own, so a client recovered the original line by accident. Adding the stuffing on the way out breaks that balance and the client ends up with `..hidden`.

A round trip over both servers, comparing the body that a client recovers against the one that was sent, with the empty line normalized away so that it measures the dot handling alone:

```
master                    ROUND TRIP INTACT      stored ..hidden   served ..hidden    recovered .hidden
this branch, POP only     ROUND TRIP CORRUPTED   stored ..hidden   served ...hidden   recovered ..hidden
this branch, both halves  ROUND TRIP INTACT      stored .hidden    served ..hidden    recovered .hidden
```

So `SMTPConnection` gained an `unstuff` that mirrors the `stuff` of the POP side, and `on_data_smtp` applies it before the message reaches the adapter. Storage now holds the message as it was written, which is what makes the stuffing on the way out correct.

Two details worth pointing out:

* The unstuffing is applied in the storage path only, not in `on_raw_data`. A relay accumulates the very same payload in `on_data_smtp` and sends it onward through `SMTPClient`, which transmits it verbatim without stuffing it again, so the relay has to keep the wire form. Undoing the stuffing earlier would have broken relaying.
* The context that it carries is three bytes rather than the two of the POP side, because the sequence being matched is `\r\n..` and one less byte than the pattern is what a boundary can hide.

The count of octets that `RETR` announces is now the size of the stored message, which is smaller than the payload that is put on the wire once the stuffing is applied. RFC 1939 treats that number as informational, the framing of the response being the terminator, and this matches what the sizes of `LIST` and `STAT` already report.

---

> [!NOTE]
> **Low Risk**
> Scoped POP3 RETR output framing fix with thorough unit tests; no auth or shared infrastructure changes.
> 
> **Overview**
> Fixes POP3 **RETR** so message bodies are framed correctly for clients instead of being truncated when a line starts with `.` or gaining an extra blank line before the end marker.
> 
> **RETR** now runs each streamed chunk through a new **`stuff`** helper that RFC 1939 byte-stuffs lines beginning with a dot (`\r\n.` → `\r\n..`), including the first body line and lines split across **`CHUNK_SIZE`** reads. A per-connection **`tail`** (last two bytes, reset at each retrieval) carries line context across chunks so stuffing still works when `\r` and `\n` straddle a read boundary.
> 
> The message terminator is sent as `.\r\n` when the body already ends with a newline, and as `\r\n.\r\n` only when it does not, avoiding an extra empty line. Stubs, **CHANGELOG**, and unit tests cover stuffed lines, empty bodies, and chunked streaming.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3e320522472822c3d15013eccf66ee705a7260f1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by joamag** [comment]
bugbot review

File: src/netius/servers/pop.py [hunk 1]
```diff
@@ -202,17 +203,44 @@ def callback(connection):
...
+
+        # stores the last bytes of the (unstuffed) body so that they may be
+        # used as the context of the next chunk, note that this is done
+        # before the stuffing as the next chunk follows the original body
+        self.tail = buffer[-2:]
+
+        # duplicates the dot of every line that starts with one, as required
+        # by the RFC 1939 specification, so that such a line is not taken
+        # for the terminator of the message by the client
+        buffer = buffer.replace(b"\r\n.", b"\r\n..")
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/servers/pop.py:238]: Accepted, the finding is correct and the branch would have corrupted the mail that arrives through SMTP. Fixed in 6bd4b042.

The mechanism is the one described: `on_raw_data` forwards the payload of the DATA command verbatim and `on_data_smtp` appends it as it is, so the stuffing of the sender survives into storage and a source line of `.hidden` is kept as `..hidden`.

Reproduced over both servers before writing anything, comparing the body that a client recovers against the one that was sent, with the empty line normalized away so that it measures the dot handling alone:

```
master                    ROUND TRIP INTACT      stored ..hidden   served ..hidden    recovered .hidden
this branch, POP only     ROUND TRIP CORRUPTED   stored ..hidden   served ...hidden   recovered ..hidden
this branch, both halves  ROUND TRIP INTACT      stored .hidden    served ..hidden    recovered .hidden
```

Worth recording that master is correct by accident rather than by design, the two mistakes cancelling each other out, which is why this only showed up once the stuffing was added on the way out.

`SMTPConnection` now has an `unstuff` that mirrors the `stuff` of the POP side, applied by `on_data_smtp` before the message reaches the adapter, so that storage holds the message as it was written.

Two details that the review did not name and that shaped the fix:

- The unstuffing is applied in the storage path only and not in `on_raw_data`. A relay accumulates the very same payload in `on_data_smtp` and sends it onward through `SMTPClient`, which transmits it verbatim without stuffing it again, so the relay has to keep the wire form. Undoing the stuffing any earlier would have broken relaying.
- The context that is carried across the chunks is three bytes rather than the two of the POP side, since the sequence being matched is `\r\n..` and a boundary can hide one byte less than the length of the pattern.

Covered by `test_unstuff`, `test_unstuff_inner`, `test_unstuff_boundary`, `test_unstuff_boundary_split`, `test_unstuff_boundary_newline` and `test_on_data_smtp_stuffed`, all of which fail against the code before the fix.

File: CHANGELOG.md [hunk 2]
```diff
@@ -36,6 +36,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Suspending the reading of a connection no longer stops its pending data from being sent - [#82](https://github.com/hivesolutions/netius/issues/82)
 * A request that arrives while the previous response is still being sent is no longer dropped - [#87](https://github.com/hivesolutions/netius/issues/87)
 * Setting up the package no longer fails when the certificate bundle cannot be downloaded
+* A retrieved message whose body has a line starting with a dot is no longer cut short - [#96](https://github.com/hivesolutions/netius/issues/96)
```
[the remark this answers is not in the document]
Comment (joamag) [reply CHANGELOG.md:39]: Not applied.

`AGENTS.md` asks for changes to be mentioned in the unreleased section, which is exactly what this entry does. Cutting a dated version section belongs to the release process rather than to a bug fix pull request, and the same file states that the version is never bumped by hand.

The unreleased section already carried non empty Added and Fixed lists before this branch existed, so the rule quoted here is not the one this repository follows.
